output/gelf/gelf_writer.go

```diff
@@ -0,0 +1,461 @@
+	zwCompressionLevel int
+	zwCompressionType  CompressType
+}
+
+// What compression type the writer should use when sending messages
+// to the graylog2 server
+type CompressType int
+
+const (
+	CompressGzip CompressType = iota
+	CompressZlib
+	NoCompress
+)
+
+// Message represents the contents of the GELF message.  It is gzipped
+// before sending. https://docs.graylog.org/docs/gelf
+type Message struct {
+	Extra     map[string]interface{} `json:"-"`
+	Full      string                 `json:"full_message"`
+	Host      string                 `json:"host"`
+	Level     int32                  `json:"level"`
+	Short     string                 `json:"short_message"`
+	Timestamp int64                  `json:"timestamp"`
+	Version   string                 `json:"version"`
+}
+
+type SimpleMessage struct {
+	Extra     map[string]interface{}
+	Host      string
+	Level     int32
+	Message   string
+	Timestamp time.Time
+}
+
+type innerMessage Message //against circular (Un)MarshalJSON
+
+// Used to control GELF chunking.  Should be less than (MTU - len(UDP
+// header)).
+//
+// TODO: generate dynamically using Path MTU Discovery?
+const (
+	DefaultChunkSize = 1420
+	chunkedHeaderLen = 12
+)
+
+var (
+	magicChunked = []byte{0x1e, 0x0f}
+)
+
+// numChunks returns the number of GELF chunks necessary to transmit
+// the given compressed buffer.
+func numChunks(b []byte, chunkSize int) int {
+	lenB := len(b)
+	if lenB <= chunkSize {
+		return 1
+	}
+	return len(b)/(chunkSize-chunkedHeaderLen) + 1
+}
+
+// NewWriter returns a new GELFWriter. This writer can be used to send the
+// output of the standard Go log functions to a central GELF server by
+// passing it to log.SetOutput()
+func NewWriter(config GELFConfig) (GELFWriter, error) {
+
+	// handle config
+	if config.Host == "" {
+		return nil, fmt.Errorf("missing host")
+	}
+
+	if config.ChunkSize == 0 {
+		config.ChunkSize = DefaultChunkSize
+	}
+
+	if config.CompressionLevel == 0 {
+		config.CompressionLevel = flate.BestSpeed
+	}
+
+	// handle http / udp writing method
+	if strings.HasPrefix(config.Host, "http") {
+		return newHTTPWriter(&config)
+	}
+
+	return newUDPWriter(&config)
+}
+
+func newHTTPWriter(config *GELFConfig) (GELFWriter, error) {
+	httpClient := &http.Client{
+		Transport: &http.Transport{},
+		Timeout:   10 * time.Second,
+	}
+
+	return HTTPWriter{
+		config:     config,
+		httpClient: httpClient,
+	}, nil
+}
+
+func newUDPWriter(config *GELFConfig) (GELFWriter, error) {
+	var err error
+
+	w := &UDPWriter{
+		config: config,
+	}
+
+	if w.hostname, err = os.Hostname(); err != nil {
+		return nil, err
+	}
+
+	if w.conn, err = net.Dial("udp", config.Host); err != nil {
+		return nil, err
+	}
+
+	return w, nil
+}
+
+func constructMessage(sm *SimpleMessage) *Message {
+
+	message := Message{
+		Extra:     sm.Extra,
+		Level:     6, // Default: Info
+		Timestamp: int64(time.Now().UnixNano() / int64(time.Millisecond) / 1000),
+		Version:   "1.1",
+	}
+
+	message.Host, _ = os.Hostname()
```

Comment:
You can panic on this err to keep the returning type simple, Assume os.Hostname() should not return err.

One more thing, the hostname can be cached if you don't want to support the hostname changes in the process life cycle.

Reply:
os.Hostname was already previously called on the creation of the Writer.
I did a global var called hostname and that hostname is fill on the creation of the Writer (And it panic if there is an error)